server/src/main/java/org/elasticsearch/ingest/IngestSourceAndMetadata.java

```diff
@@ -0,0 +1,613 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.index.VersionType;
+
+import java.time.ZonedDateTime;
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * Map containing ingest source and metadata.
+ *
+ * The Metadata values in {@link IngestDocument.Metadata} are validated when put in the map.
+ * _index, _id and _routing must be a String or null
+ * _version_type must be a lower case VersionType or null
+ * _version must be representable as a long without loss of precision or null
+ * _dyanmic_templates must be a map
+ * _if_seq_no must be a long or null
+ * _if_primary_term must be a long or null
+ *
+ * The map is expected to be used by processors, server code should the typed getter and setters where possible.
+ */
+class IngestSourceAndMetadata extends AbstractMap<String, Object> {
+    protected final ZonedDateTime timestamp;
+
+    /**
+     * map of key to validating function. Should throw {@link IllegalArgumentException} on invalid value, otherwise return identity
```

Comment:
Why make it return anything? If validators throw IAE, then they could just be Consumers?

Reply:
Changed to BiConsumers